src/treetensornetworks/projttns/projttn.jl

```diff
@@ -20,38 +21,83 @@ function set_nsite(P::ProjTTN, nsite)
 end
 
 function shift_position(P::ProjTTN, pos)
-  return ProjTTN(pos, P.H, P.environments)
+  return ProjTTN(pos, operator(P), environments(P))
 end
 
-function make_environment!(P::ProjTTN{V}, psi::TTN{V}, e::NamedEdge{V})::ITensor where {V}
+set_environments(p::ProjTTN, environments) = ProjTTN(pos(p), operator(p), environments)
+function set_environment(p::ProjTTN, edge, env)
+  newenvs = merge(environments(p), Dictionary((edge,), (env,)))
+  return ProjTTN(pos(p), operator(p), newenvs)
+end
+set_environment!(p::ProjTTN, edge, env) = set!(environments(p), edge, env)
+
+function make_environment!(
+  P::ProjTTN{V}, state::AbstractTTN{V}, e::NamedEdge{V}
+)::ITensor where {V}
   # invalidate environment for opposite edge direction if necessary
-  reverse(e) ∈ incident_edges(P) || unset!(P.environments, reverse(e))
+  reverse(e) ∈ incident_edges(P) || unset!(environments(P), reverse(e))
   # do nothing if valid environment already present
-  if haskey(P.environments, e)
+  if haskey(environments(P), e)
     env = environment(P, e)
   else
     if is_leaf(underlying_graph(P), src(e))
       # leaves are easy
-      env = psi[src(e)] * P.H[src(e)] * dag(prime(psi[src(e)]))
+      env = state[src(e)] * operator(H)[src(e)] * dag(prime(state[src(e)]))
     else
       # construct by contracting neighbors
       neighbor_envs = ITensor[]
       for n in setdiff(neighbors(underlying_graph(P), src(e)), [dst(e)])
-        push!(neighbor_envs, make_environment!(P, psi, edgetype(P)(n, src(e))))
+        push!(neighbor_envs, make_environment!(P, state, edgetype(P)(n, src(e))))
       end
       # manually heuristic for contraction order: two environments, site tensors, then
       # other environments
       frst, scnd, rst = _separate_first_two(neighbor_envs)
-      itensor_map = vcat(psi[src(e)], frst, scnd, P.H[src(e)], dag(prime(psi[src(e)])), rst)
+      itensor_map = vcat(
+        state[src(e)], frst, scnd, operator(H)[src(e)], dag(prime(state[src(e)])), rst
+      )
       # TODO: actually use optimal contraction sequence here
       env = reduce(*, itensor_map)
     end
-    # cache
-    set!(P.environments, e, env)
+    set!(environments(P), e, env)
   end
   @assert(
-    hascommoninds(environment(P, e), psi[src(e)]),
+    hascommoninds(environment(P, e), state[src(e)]),
     "Something went wrong, probably re-orthogonalized this edge in the same direction twice!"
   )
   return env
 end
+
+function make_environment(
+  P::ProjTTN{V}, state::AbstractTTN{V}, e::NamedEdge{V}
+)::ProjTTN{V} where {V}
```

Comment:
```suggestion
function make_environment(
  P::ProjTTN, state::AbstractTTN, e::NamedEdge
)
```
Since `make_environment` and `make_environment!` are both pretty complicated, we should keep just one of them. I would prefer to just keep `make_environment` for now and then make an in-place version if needed for performance.